### DISCUSSION TITLE
httpx.Response.stream does not detect inital EOF

### DISCUSSION THREAD

**Discussion by D1-3105** [encode/httpx#3415] [Potential Issue] [answered]
Version: 0.26.0
Steps to reproduce:
0. Create httpx.AsyncClient instance
1. Send POST request to some web-server which responds with EOF as body with some status code different to 204 (I used webhook.site)
2. Call `async for _ in response.stream`
3. The httpx stream is deadlocked, while underlying tcp connection is closed.

Full snippet:
```
import asyncio
import sys
import httpx
import time

last_timer = time.perf_counter()
async def call_me():
    client = httpx.AsyncClient()

    last_status_code = None
    global last_timer
    while last_status_code != 429:
        response = await client.post("https://webhook.site/7c7048f3-44ac-4cac-8450-96031dd41a0a", timeout=httpx.Timeout(5))
        try:
            last_status_code = response.status_code
            print(f"{last_status_code=}")
            print(f"{response.stream._stream._httpcore_stream._closed=}")
            async for i in response.stream:
                print(i, flush=True)
        except Exception as e:
            print(e, flush=True)
        finally:
            await response.aclose()

try:
    asyncio.run(call_me())
finally:
    print(time.perf_counter() - last_timer)
```

My output:
```
$ python3 test.py
last_status_code=200
response.stream._stream._httpcore_stream._closed=True
^C18.568296384997666
```
As you can see, the httpx stream was waiting for nothing.

**Comment by lovelydinosaur** [comment, marked as answer]
You're using `client.post(...)` so the entire response has already been streamed and is available in `response.content`.
Use `with client.stream(...) as response` for streaming responses.

Tho yes, there's some possible tighter API constraints we could put in place there. 🤔

**Reply by D1-3105** [reply]
I feel if you allow devs to iter through something than you have to raise `StopIteration`/`StopAsyncIteration` if the iterator exceeded. Also, if there is a content in response body - `.stream` method raises `StopIteration` as it is expected. I caught a few deadlocks because of this case (`httpx.Timeout` does not do a thing, as you can see). 
Nevertheless, thanks for the proper way to iter streams, forgot about this func.

**Reply by lovelydinosaur** [reply]
> Send POST request to some web-server which responds with EOF as body with some status code different to 204 

Are you able to demo a little server or similar to test this against?
I'm only seeing properly formed 200 responses against that particular URL.

**Reply by D1-3105** [reply]
> > Send POST request to some web-server which responds with EOF as body with some status code different to 204
> 
> Are you able to demo a little server or similar to test this against? I'm only seeing properly formed 200 responses against that particular URL.

Could you please provide your output and version, I'd like to test with your environment.
